python/pyspark/sql/session.py

```diff
@@ -501,18 +504,20 @@ def getOrCreate(self) -> "SparkSession":
                             SparkContext._active_spark_context is None
                             and SparkSession._instantiatedSession is None
                         ):
-                            if is_api_mode_connect:
-                                url = opts.get("spark.master", os.environ.get("MASTER"))
-                            else:
-                                url = opts.get("spark.remote", os.environ.get("SPARK_REMOTE"))
+                            url = opts.get("spark.remote", os.environ.get("SPARK_REMOTE"))
+
+                            if url is None and is_api_mode_connect:
+                                url = opts.get("spark.master", os.environ.get("MASTER", "local"))
 
                             if url is None:
                                 raise PySparkRuntimeError(
                                     errorClass="CONNECT_URL_NOT_SET",
                                     messageParameters={},
                                 )
 
-                            if url.startswith("local") or is_api_mode_connect:
+                            if url.startswith("local") or (
+                                is_api_mode_connect and not url.startswith("sc://")
```

Comment:
What does this condition mean? connect api mode with non-local and non-sc master?

Reply:
Yes. It will try to start a cluster in connect mode, like yarn or k8s.

Reply:
@HyukjinKwon can you help to provide more context here? I think for all resource managers (local, master, yarn, k8s) we need to start the temp connect server, right?

Reply:
yes. I will followup.